hack/reduce-snapshot/reduce-snapshot.sh

```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env bash
+set -eux
+
+echo "Single Component mode? ${SINGLE_COMPONENT}"
+if [ "${SINGLE_COMPONENT}" == "true" ]; then
+
+  CR_NAMESPACE_ARG=
+  if [ "${CUSTOM_RESOURCE_NAMESPACE}" != ""]; then
+    CR_NAMESPACE_ARG="-n ${CUSTOM_RESOURCE_NAMESPACE}"
+  fi
+  SNAPSHOT_CREATION_TYPE=$(kubectl get "$CUSTOM_RESOURCE" ${CR_NAMESPACE_ARG:+$CR_NAMESPACE_ARG} -ojson \
+      | jq -r '.metadata.labels."test.appstudio.openshift.io/type" // ""')
+  SNAPSHOT_CREATION_COMPONENT=$(kubectl get "$CUSTOM_RESOURCE" ${CR_NAMESPACE_ARG:+$CR_NAMESPACE_ARG} -ojson \
+      | jq -r '.metadata.labels."appstudio.openshift.io/component" // ""')
+
+  echo "SNAPSHOT_CREATION_TYPE: ${SNAPSHOT_CREATION_TYPE}"
+  echo "SNAPSHOT_CREATION_COMPONENT: ${SNAPSHOT_CREATION_COMPONENT}"
+  if [ "${SNAPSHOT_CREATION_TYPE}" == "component" ] && [ "${SNAPSHOT_CREATION_COMPONENT}" != "" ]; then
+    echo "Single Component mode is ${SINGLE_COMPONENT} and Snapshot type is component"
+    # verify if in json form
+    set +e
+    echo "${SNAPSHOT}" | jq -e > /dev/null 2>&1
+    CHECK_JSON=$?
+    set -e
+    if [ "${CHECK_JSON}" != "0" ]; then
+      # try to load as file first
+      set +e
+      jq -e < "${SNAPSHOT}" > /dev/null 2>&1
+      CHECK_JSON_FILE=$?
+      set -e
+      if [ "${CHECK_JSON_FILE}" != "0" ]; then
+        echo "Cannot load snapshot from json string or file"
+        exit 1
+      fi
+      SNAPSHOT=$(cat "${SNAPSHOT}")
+    fi
+    REDUCED_SNAPSHOT=$(echo "${SNAPSHOT}" | jq --arg component "${SNAPSHOT_CREATION_COMPONENT}" \
+    'del(.components[] | select(.name != $component))')
+
+    ## make sure we still have 1 component
+    COMPONENT_COUNT=$(echo "$REDUCED_SNAPSHOT" | jq -r '[ .components[] ] | length')
+    echo "COMPONENT_COUNT: ${COMPONENT_COUNT}"
+    if [ "${COMPONENT_COUNT}" != "1" ] ; then
+      echo "Error: Reduced Snapshot has ${COMPONENT_COUNT} components. It should contain 1"
+      echo "       Verify that the Snapshot contains the built component: ${SNAPSHOT_CREATION_COMPONENT}"
+      exit 1
+    fi
+
+    echo "Reducing Snapshot to:"
+    echo "$REDUCED_SNAPSHOT" | jq .
+    SNAPSHOT=$(echo "$REDUCED_SNAPSHOT" | tr -d ' ' | tr -d '\n')
+    echo "$SNAPSHOT" > "${SNAPSHOT_PATH}"
```

Comment:
Might be dangerous if any of the keys or values in the JSON contain the space character, why not something like this:

```suggestion
    echo "$SNAPSHOT" | jq -c . > "${SNAPSHOT_PATH}"
```

Reply:
fixed. thanks

Reply:
I see this now:
```
echo "$REDUCED_SNAPSHOT" | jq .
SNAPSHOT=$(echo "$REDUCED_SNAPSHOT" | tr -d ' ' | tr -d '\n')
echo "$SNAPSHOT" | jq -c . > "${SNAPSHOT_PATH}"
```

I think @zregvart meant this:

```
# for displaying
echo "$REDUCED_SNAPSHOT" | jq .
# for saving
echo "$REDUCED_SNAPSHOT" | jq -c . > "${SNAPSHOT_PATH}"
```

Honestly, I would just do something like this:

```
echo "$REDUCED_SNAPSHOT" | jq '.' | tee "${SNAPSHOT_PATH}"
```

I don't really think we need to save the reduced snapshot as a compacted json.

Reply:
ok adjusted. but I seem to remember an issue whereby I needed to remove the line feeds...